nio/client/async_client.py

```diff
@@ -826,6 +835,15 @@ async def register(self, username, password, device_name=""):
 
         return await self._send(RegisterResponse, method, path, data)
 
+    async def discovery_info(self) -> Union[LoginInfoResponse, LoginInfoError]:
+        """Get discovery information about current `AsyncClient.homeserver`.
+
+        Returns either a `DiscoveryInfoResponse` if the request was successful
```

Comment:
I'm a bit confused on how you're using this, to call this method you need an `AsyncClient` instance, but to create an `AsyncClient` instance you need a homeserver URL in the constructor.

The spec clearly describes the steps that should be taken to do autodiscovery.

> Extract the server name from the user's Matrix ID by splitting the Matrix ID at the first colon.

This step seems to be completely missing.

Wouldn't a separate `AsyncClient` constructor be the way to go for this?

Reply:
This step assumes the user will sign in with a public server's user ID + password combo, which doesn't make sense: what if they only pass a username? What if their user ID is `@something:localhost` (default Synapse config, notice the lack of port in the hostname)? What if they use SSO? etc... To even know which login flow the server supports and request a user ID (or not), the first step is to know the server's "real" URL using .well-known, which is the reverse of what everyone strangely seems to agree on.

The other way is to have a default hardcoded server, which encourages centralization and offers a frustrating extra clicks back-and-forth user experience to anyone outside of that server, which is what Element does.

This is the procedure I currently use for Mirage:
- Ask the user to pick a server
- [Get its real URL and supported login flows](https://github.com/mirukana/mirage/blob/49274fc/src/backend/backend.py#L136)
- For `m.login.password`, ask a username/ID + password, [create a client and do password login](https://github.com/mirukana/mirage/blob/49274fc/src/backend/backend.py#L160)
- For `m.login.sso`, [start the SSO process and create a client](https://github.com/mirukana/mirage/blob/49274fc/src/backend/backend.py#L185) 

Of course, this is all my opinionated way of doing things, so I've opted to only add the basic API into nio and others can use them in whichever way they like. Would you prefer to have some kind of abstraction layer/automatic constructor before merging this?

Reply:
> This step assumes the user will sign in with a public server's user ID + password combo, which doesn't make sense: what if they only pass a username?

I don't think a password is required, only that the user inputs the full `user_id`, just like some mail services require the user + domain for login.

> What if their user ID is `@something:localhost` (default Synapse config, notice the lack of port in the hostname)?

I don't think Synapse hosts the well-known, you'll need a well-known file at `localhost:433` just like you would need to have it usually.

> What if they use SSO? etc... To even know which login flow the server supports and request a user ID (or not), the first step is to know the server's "real" URL using .well-known, which is the reverse of what everyone strangely seems to agree on.

Yes but to know the real server URL, as per spec, you use the full user id, while in your example you use the server URL to get the server URL?

> The other way is to have a default hardcoded server, which encourages centralization and offers a frustrating extra clicks back-and-forth user experience to anyone outside of that server, which is what Element does.

I don't think you need to hard-code anything, Fractal supports this and it doesn't hard-code it. 

> Would you prefer to have some kind of abstraction layer/automatic constructor before merging this?

The way I understand this feature and have seen it in other clients is as follows:
1. Offer the user to enter the user name
2. If it's a full user id discover the server
3.  If it's not a full user id or the discovery fails, request a server to be input.
4.  Once you have the server you can check for the login methods.

A nice way to present this in API form would be a separate constructor I would think.

For exapmle

```python
try:
    client = AsyncClient.with_auto_discovery(user_id, "store_path"...)
except DiscoveryError(e):
    # present some error here and request a server
    pass
```
Wouldn't this work for Mirage?

Reply:
> Yes but to know the real server URL, as per spec, you use the full user id, while in your example you use the server URL to get the server URL?

I take a domain, e.g. "https://matrix.org", and try to retrieve its .well-known, which gives me e.g. "https://matrix-client.matrix.org". If getting the well-known fails, I assume it's a direct server URL, and if getting the login flows of that fails, I assume it's not a valid URL at all. 

> I don't think you need to hard-code anything, Fractal supports this and it doesn't hard-code it.

I've tested the latest Fractal release 4.2.2, it asks for a server first, then asks for user/password (and doesn't support SSO).

The main problem is still this: for some login flows, the user doesn't enter a user ID at all. You can try it with Element: change the server to "https://mozilla.modular.im", it will show you a button to start the SSO procedure which is out of the client's hands, then you're in with an automatically assigned user ID.

Reply:
> I take a domain, e.g. "https://matrix.org", and try to retrieve its .well-known, which gives me e.g. "https://matrix-client.matrix.org". If getting the well-known fails, I assume it's a direct server URL, and if getting the login flows of that fails, I assume it's not a valid URL at all.

I guess it's useful for this use-case, which is completely orthogonal to the server discovery via user id. I'm not really sure how many servers require discovery after you already input the main domain but I'm fine with this patch. 

Could you just expand the docs with an example, it's not quite clear how this should be used considering that it's for an unexpected use-case given familiarity with the spec.

Reply:
I expanded the docs, is it good enough?